# Project submission
**Due Friday May 16th before class.** Counts for 25% of the final course grade.

You should address all the questions relevant to your project.
You will not be graded based on the values of the model performance, but on whether or not you have applied the right methodology: formulated the business model, translated it into a right machine learning approach, analyzed your data, prepared it for modeling, applied at least 5 different machine learning algorithms, as well as neural networks, used cross validation for model tuning, justified your tuning metric, set up the proper machine learning pipeline without data leakage, evaluated your model using all the relevant metrics, interpreted your model and justified all your decisions.

If you have tried different approaches, please include them all, and not just the best one.
If doing some feature engineering has improved your model, also please include all of the steps, not just the most successful ones.

You should submit the notebook with the code, output and explanations. The notebook should be executable and comprehensible.

The points will be deducted for the following reasons:
- data leakage
- unjustified decisions (no discussion on: choice of metric for optimization, blind removal of features, blind removal of outliers...)
- notebook not comprehensible
- notebook with incomplete output
- notebook not executable
- blind copy pasting from ChatGPT, if the copied code is not suitable for the task
- writing your own code (or copy pasting them from outside source) for simple functions that we covered and that already exist in `sklearn` (train test split, plain grid search, encoding of categorical variables,...), as this leads to:
    - convoluted code prone to bugs
    - code that is hard to understand and review
    - waste of data scientist's time if ready-made simple functions exist

Additional points will be awarded for trying and testing different relevant approaches, from exploratory data analysis, to feature engineering, to modeling and evaluation.

There should be one submission per group, but team member evaluation can be submitted per person. If not submitted, the default is that all the team members have contributed equally to the project and should get the same grade.

### Group number:
### Student IDs:
### Project name:

## What business problem are you solving?
- Please state clearly what business problem are you solving. (one sentence)
- Elaborate why is this a relevant problem, and what can you do with the model output to create business value, i.e., how is the model output actionable. (2-3 paragraphs)

## What is the machine learning problem that you are solving?
- Please state clearly what is the ML problem. 
- If applicable state your target.

## Data exploration and preparation 

- How many data instances do you have?
- Do you have duplicates?
- How many features? What type are they?
- If they are categorical, what categories they have, what is their frequency?
- If they are numerical, what is their distribution?
- Do you have outliers, and do you need to do anything about them?
- What is the distribution of the target variable?
- If you have a target, you can also check the relationship between the target and the variables.
- Do you have missing data? If yes, how are you going to handle it?
- Can you use the features in their original form, or do you need to alter them in some way?
- What have you learned about your data? Is there anything that can help you in feature engineering or modeling?


## Feature engineering
Creating good features is probably the most important step in the machine learning process. 
This might involve doing:
- transformations
- aggregating over data points or over time and space, or finding differences (for example: differences between two monthly bills, time difference between two contacts with the client) 
- creating dummy (binary) variables
- discretization

Business insight is very relevant in this process. If it is possible you can also find additional relevant data.

## Modeling
You should implement AT LEAST FIVE approaches we covered, and tune of at least two hyperparameters of each approach.
Do not forget that you should split your data.
You should do model selection and tuning using cross validation on the train set, avoiding data leakage.
Explain and justify what is the metric you are using for model selection and tuning. If your data is imbalanced, consider using techniques for data balancing.

Separately, you should train a neural network. Visualize the training and validation loss. Discuss the network performance

In model selection, make sure when you compare different models and approaches that you compare them on the same dataset, though different transformations could be applied to the comparison dataset.

## Model evaluation

After selecting your final model, which could be a compromise of performance, interpretability and complexity, you should evaluate its performance on the test set. 
You might have tuned your model using a certain metric, but now you should describe the model performance using all relevant metrics. 
If you have some business insight, why a certain metric is relevant, you should explain it. 
Construct a suitable baseline to benchmark your result and to put them in the context.
Discuss your results, do they seem good enough to be used in practice? If not, what should be improved. Discuss what type of errors is your model making.


## Model interpretation

Use at least two different techniques for model interpretability. Discuss what are the most important features of your model, and how they impact the model performance. Pick a few examples of errors that your model is making, and check which features lead to thess errors.

In [130]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_colwidth', None)

In [131]:
df = pd.read_csv('pr13_stocks (1).csv', index_col=0)
df.head()

Date  Dividends  Stock Splits  \
0  2021-01-25 00:00:00-05:00        0.0           0.0   
1  2019-09-12 00:00:00-04:00        0.0           0.0   
2  2015-12-29 00:00:00-05:00        0.0           0.0   
3  2014-06-13 00:00:00-04:00        0.0           0.0   
4  2017-10-06 00:00:00-04:00        0.0           0.0   

                Brand_Name Ticker    Industry_Tag      Country      Volume  \
0                    crocs   CROX        footwear          usa   1102500.0   
1                   target    TGT          retail          usa   3185700.0   
2                 unilever     UL  consumer goods  netherlands   1278700.0   
3                      amd    AMD      technology          usa  17734600.0   
4  the walt disney company    DIS   entertainment          usa   4360200.0   

         Open        High        Low       Close  
0   73.180000   74.750000  71.050003   73.910004  
1  100.810594  101.077753  99.999911  100.359192  
2   33.785507   33.970634  33.700657   33.908924  
3    4.360000    4.390000   4.240000    4.280000  
4   96.426167   96.734885  95.837672   96.541939

In [132]:
df['Date'] = pd.to_datetime(df['Date'], utc=True)

df

Date  Dividends  Stock Splits  \
0     2021-01-25 05:00:00+00:00        0.0           0.0   
1     2019-09-12 04:00:00+00:00        0.0           0.0   
2     2015-12-29 05:00:00+00:00        0.0           0.0   
3     2014-06-13 04:00:00+00:00        0.0           0.0   
4     2017-10-06 04:00:00+00:00        0.0           0.0   
...                         ...        ...           ...   
99995 2004-06-18 04:00:00+00:00        0.0           0.0   
99996 2022-06-24 04:00:00+00:00        0.0           0.0   
99997 2017-09-06 04:00:00+00:00        0.0           0.0   
99998 2021-06-04 04:00:00+00:00        0.0           0.0   
99999 2011-08-11 04:00:00+00:00        0.0           0.0   

                    Brand_Name Ticker     Industry_Tag      Country  \
0                        crocs   CROX         footwear          usa   
1                       target    TGT           retail          usa   
2                     unilever     UL   consumer goods  netherlands   
3                          amd    AMD       technology          usa   
4      the walt disney company    DIS    entertainment          usa   
...                        ...    ...              ...          ...   
99995                    fedex    FDX        logistics          usa   
99996          hershey company    HSY  food & beverage          usa   
99997                 unilever     UL   consumer goods  netherlands   
99998                   amazon   AMZN       e-commerce          usa   
99999       salesforce / slack    CRM       technology          usa   

           Volume        Open        High         Low       Close  
0       1102500.0   73.180000   74.750000   71.050003   73.910004  
1       3185700.0  100.810594  101.077753   99.999911  100.359192  
2       1278700.0   33.785507   33.970634   33.700657   33.908924  
3      17734600.0    4.360000    4.390000    4.240000    4.280000  
4       4360200.0   96.426167   96.734885   95.837672   96.541939  
...           ...         ...         ...         ...         ...  
99995    957800.0   66.332358   66.936849   65.991798   66.604805  
99996   1154300.0  213.073770  216.005978  212.311397  215.966888  
99997   1079900.0   47.701035   47.904677   47.546265   47.880238  
99998  44994000.0  160.600006  161.050003  159.940506  160.311005  
99999  10658000.0   32.407501   34.180000   32.047501   33.637501  

[100000 rows x 12 columns]

In [133]:
df.isna().sum()

Date             16
Dividends        59
Stock Splits    458
Brand_Name      225
Ticker          175
Industry_Tag     34
Country         100
Volume          852
Open            803
High            664
Low             681
Close           194
dtype: int64

In [134]:
#drop the 16 rows without any date data
df = df.dropna(subset=['Date'])

In [135]:
df

Date  Dividends  Stock Splits  \
0     2021-01-25 05:00:00+00:00        0.0           0.0   
1     2019-09-12 04:00:00+00:00        0.0           0.0   
2     2015-12-29 05:00:00+00:00        0.0           0.0   
3     2014-06-13 04:00:00+00:00        0.0           0.0   
4     2017-10-06 04:00:00+00:00        0.0           0.0   
...                         ...        ...           ...   
99995 2004-06-18 04:00:00+00:00        0.0           0.0   
99996 2022-06-24 04:00:00+00:00        0.0           0.0   
99997 2017-09-06 04:00:00+00:00        0.0           0.0   
99998 2021-06-04 04:00:00+00:00        0.0           0.0   
99999 2011-08-11 04:00:00+00:00        0.0           0.0   

                    Brand_Name Ticker     Industry_Tag      Country  \
0                        crocs   CROX         footwear          usa   
1                       target    TGT           retail          usa   
2                     unilever     UL   consumer goods  netherlands   
3                          amd    AMD       technology          usa   
4      the walt disney company    DIS    entertainment          usa   
...                        ...    ...              ...          ...   
99995                    fedex    FDX        logistics          usa   
99996          hershey company    HSY  food & beverage          usa   
99997                 unilever     UL   consumer goods  netherlands   
99998                   amazon   AMZN       e-commerce          usa   
99999       salesforce / slack    CRM       technology          usa   

           Volume        Open        High         Low       Close  
0       1102500.0   73.180000   74.750000   71.050003   73.910004  
1       3185700.0  100.810594  101.077753   99.999911  100.359192  
2       1278700.0   33.785507   33.970634   33.700657   33.908924  
3      17734600.0    4.360000    4.390000    4.240000    4.280000  
4       4360200.0   96.426167   96.734885   95.837672   96.541939  
...           ...         ...         ...         ...         ...  
99995    957800.0   66.332358   66.936849   65.991798   66.604805  
99996   1154300.0  213.073770  216.005978  212.311397  215.966888  
99997   1079900.0   47.701035   47.904677   47.546265   47.880238  
99998  44994000.0  160.600006  161.050003  159.940506  160.311005  
99999  10658000.0   32.407501   34.180000   32.047501   33.637501  

[99984 rows x 12 columns]

In [136]:
df[df['Brand_Name'].isna() & df['Ticker'].isna()]

Date  Dividends  Stock Splits Brand_Name Ticker  \
29955 2000-11-16 05:00:00+00:00        0.0           0.0        NaN    NaN   

      Industry_Tag Country      Volume      Open      High       Low     Close  
29955      apparel     usa  17245350.0  4.804951  5.124243  4.804951  5.038579

In [137]:
#drop the row without both Brand_name and Ticker as we cannot say which company this entry refers to
df = df[~(df['Brand_Name'].isna() & df['Ticker'].isna())].copy()

In [138]:
df

Date  Dividends  Stock Splits  \
0     2021-01-25 05:00:00+00:00        0.0           0.0   
1     2019-09-12 04:00:00+00:00        0.0           0.0   
2     2015-12-29 05:00:00+00:00        0.0           0.0   
3     2014-06-13 04:00:00+00:00        0.0           0.0   
4     2017-10-06 04:00:00+00:00        0.0           0.0   
...                         ...        ...           ...   
99995 2004-06-18 04:00:00+00:00        0.0           0.0   
99996 2022-06-24 04:00:00+00:00        0.0           0.0   
99997 2017-09-06 04:00:00+00:00        0.0           0.0   
99998 2021-06-04 04:00:00+00:00        0.0           0.0   
99999 2011-08-11 04:00:00+00:00        0.0           0.0   

                    Brand_Name Ticker     Industry_Tag      Country  \
0                        crocs   CROX         footwear          usa   
1                       target    TGT           retail          usa   
2                     unilever     UL   consumer goods  netherlands   
3                          amd    AMD       technology          usa   
4      the walt disney company    DIS    entertainment          usa   
...                        ...    ...              ...          ...   
99995                    fedex    FDX        logistics          usa   
99996          hershey company    HSY  food & beverage          usa   
99997                 unilever     UL   consumer goods  netherlands   
99998                   amazon   AMZN       e-commerce          usa   
99999       salesforce / slack    CRM       technology          usa   

           Volume        Open        High         Low       Close  
0       1102500.0   73.180000   74.750000   71.050003   73.910004  
1       3185700.0  100.810594  101.077753   99.999911  100.359192  
2       1278700.0   33.785507   33.970634   33.700657   33.908924  
3      17734600.0    4.360000    4.390000    4.240000    4.280000  
4       4360200.0   96.426167   96.734885   95.837672   96.541939  
...           ...         ...         ...         ...         ...  
99995    957800.0   66.332358   66.936849   65.991798   66.604805  
99996   1154300.0  213.073770  216.005978  212.311397  215.966888  
99997   1079900.0   47.701035   47.904677   47.546265   47.880238  
99998  44994000.0  160.600006  161.050003  159.940506  160.311005  
99999  10658000.0   32.407501   34.180000   32.047501   33.637501  

[99983 rows x 12 columns]

In [139]:
df['Brand_Name'] = df['Brand_Name'].str.lower()
df['Ticker'] = df['Ticker'].str.upper()

brand_to_ticker = df.dropna(subset=['Brand_Name', 'Ticker'])\
                        .drop_duplicates(subset=['Brand_Name'])\
                        .set_index('Brand_Name')['Ticker'].to_dict()

ticker_to_brand = df.dropna(subset=['Brand_Name', 'Ticker'])\
                        .drop_duplicates(subset=['Ticker'])\
                        .set_index('Ticker')['Brand_Name'].to_dict()

    # Fill missing Ticker using Brand_Name
df.loc[df['Ticker'].isna() & df['Brand_Name'].notna(), 'Ticker'] = (
        df.loc[df['Ticker'].isna() & df['Brand_Name'].notna(), 'Brand_Name']
        .map(brand_to_ticker)
    )

    # Fill missing Brand_Name using Ticker
df.loc[df['Brand_Name'].isna() & df['Ticker'].notna(), 'Brand_Name'] = (
    df.loc[df['Brand_Name'].isna() & df['Ticker'].notna(), 'Ticker']
    .map(ticker_to_brand))

In [140]:
df.isna().sum() 

Date              0
Dividends        59
Stock Splits    458
Brand_Name        0
Ticker            0
Industry_Tag     34
Country         100
Volume          852
Open            803
High            664
Low             681
Close           194
dtype: int64

In [141]:
#As dividend payments and stock splits are sparse, event-based features, missing values here are likely because 
#no such event has occured. Therefore, missing values are filled with 0. Also, the occurence of missing values 
#is relatively low here(59 and 458, respectively) compared to the overall dataset size(~100,000)
df['Dividends'] = df['Dividends'].fillna(0.0)
df['Stock Splits'] = df['Stock Splits'].fillna(0.0)

In [142]:
df.isna().sum()

Date              0
Dividends         0
Stock Splits      0
Brand_Name        0
Ticker            0
Industry_Tag     34
Country         100
Volume          852
Open            803
High            664
Low             681
Close           194
dtype: int64

In [143]:
df['Industry_Tag'].value_counts()

Industry_Tag
technology            21573
retail                 8600
automotive             8047
finance                7146
apparel                6708
consumer goods         6378
food & beverage        6371
entertainment          4095
food                   3836
footwear               3748
aviation               3583
gaming                 3575
hospitality            3310
e-commerce             2797
healthcare             2153
logistics              2106
manufacturing          2105
luxury goods           1597
financial services      726
music                   483
social media            396
fitness                 376
cryptocurrency          220
techcology                1
fooT                      1
technRlogy                1
tecanology                1
tecbnology                1
financial Tervices        1
fgnance                   1
retasl                    1
e-commVrce                1
healthcarE                1
axparel                   1
ghming                    1
tRchnol

In [144]:
#Imputing industry tag values with the most frequent value for each brand

df['Industry_Tag'] = df['Industry_Tag'].str.strip().str.lower()
industry_map = (
        df.dropna(subset=['Industry_Tag'])
          .groupby('Brand_Name')['Industry_Tag']
          .agg(lambda x: x.mode()[0]) 
          .to_dict()
    )

df['Industry_Tag'] = df.apply(
        lambda row: industry_map[row['Brand_Name']]
        if pd.isna(row['Industry_Tag']) or row['Industry_Tag'] != industry_map.get(row['Brand_Name'])
        else row['Industry_Tag'],
        axis=1
    )

In [145]:
df['Industry_Tag'].value_counts()

Industry_Tag
technology            21587
retail                 8606
automotive             8054
finance                7150
apparel                6710
consumer goods         6381
food & beverage        6372
entertainment          4096
food                   3838
footwear               3748
aviation               3584
gaming                 3577
hospitality            3310
e-commerce             2803
healthcare             2155
logistics              2106
manufacturing          2106
luxury goods           1597
financial services      728
music                   483
social media            396
fitness                 376
cryptocurrency          220
Name: count, dtype: int64

In [146]:
#same for country

df['Country'] = df['Country'].str.strip()

country_map = (
        df.dropna(subset=['Country'])
          .groupby('Brand_Name')['Country']
          .agg(lambda x: x.mode()[0]) 
          .to_dict()
    )

df['Country'] = df.apply(
        lambda row: country_map[row['Brand_Name']]
        if pd.isna(row['Country']) or row['Country'] != country_map.get(row['Brand_Name'])
        else row['Country'],
        axis=1
    )

In [147]:
print(df['Country'].value_counts())
print(df.isna().sum())
print(f"Number of duplicates: {df.duplicated().sum()}")

Country
usa            78572
japan           6413
germany         5082
netherlands     4263
france          2823
switzerland     2132
canada           698
Name: count, dtype: int64
Date              0
Dividends         0
Stock Splits      0
Brand_Name        0
Ticker            0
Industry_Tag      0
Country           0
Volume          852
Open            803
High            664
Low             681
Close           194
dtype: int64
Number of duplicates: 0


In [148]:
print(df.shape)
df = df.drop(columns=['Volume', 'Open', 'High', 'Low','Brand_Name'])
df = df.dropna(subset=['Close'])
print(df.shape)

(99983, 12)
(99789, 7)


Delete Open, High and Low since this data is same day date and not lagged --> otherwise data leakage
Delete Brand_Name because it contains the same info as Ticker
Delete 194 with no Data since it is only 194/100k values and is spread between dates/companies/strategies

In [149]:
print(df.columns)
df[df['Ticker'] == 'FL'][:20]

Index(['Date', 'Dividends', 'Stock Splits', 'Ticker', 'Industry_Tag',
       'Country', 'Close'],
      dtype='object')


Date  Dividends  Stock Splits Ticker Industry_Tag  \
68  2019-07-23 04:00:00+00:00        0.0           0.0     FL     footwear   
106 2001-01-24 05:00:00+00:00        0.0           0.0     FL     footwear   
109 2014-09-23 04:00:00+00:00        0.0           0.0     FL     footwear   
180 2013-08-13 04:00:00+00:00        0.0           0.0     FL     footwear   
242 2023-09-08 04:00:00+00:00        0.0           0.0     FL     footwear   
334 2013-08-16 04:00:00+00:00        0.0           0.0     FL     footwear   
348 2011-05-20 04:00:00+00:00        0.0           0.0     FL     footwear   
359 2002-05-01 04:00:00+00:00        0.0           0.0     FL     footwear   
361 2021-12-15 05:00:00+00:00        0.0           0.0     FL     footwear   
378 2001-07-05 04:00:00+00:00        0.0           0.0     FL     footwear   
380 2019-09-30 04:00:00+00:00        0.0           0.0     FL     footwear   
390 2012-05-21 04:00:00+00:00        0.0           0.0     FL     footwear   
563 2003-08-26 04:00:00+00:00        0.0           0.0     FL     footwear   
646 2003-11-25 05:00:00+00:00        0.0           0.0     FL     footwear   
685 2015-04-23 04:00:00+00:00        0.0           0.0     FL     footwear   
687 2017-12-08 05:00:00+00:00        0.0           0.0     FL     footwear   
717 2016-05-19 04:00:00+00:00        0.0           0.0     FL     footwear   
796 2021-05-26 04:00:00+00:00        0.0           0.0     FL     footwear   
809 2007-04-04 04:00:00+00:00        0.0           0.0     FL     footwear   
867 2010-11-26 05:00:00+00:00        0.0           0.0     FL     footwear   

    Country      Close  
68      usa  37.268436  
106     usa   6.607511  
109     usa  43.721626  
180     usa  26.613600  
242     usa  18.139999  
334     usa  26.392712  
348     usa  18.100170  
359     usa   8.785510  
361     usa  39.288742  
378     usa   8.904234  
380     usa  37.459381  
390     usa  22.538576  
563     usa   9.211951  
646     usa  12.526238  
685     usa  47.782459  
687     usa  37.146862  
717     usa  46.803654  
796     usa  58.207123  
809     usa  14.210990  
867     usa  13.230009

In [150]:
df['Date'] = pd.to_datetime(df['Date']) ### was machen wir mit den Stunden???

# Time-based features
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month
df['Day'] = df['Date'].dt.day
df['DayOfWeek'] = df['Date'].dt.dayofweek  # 0 = Monday, 6 = Sunday

# Running day number
df['Day_Number'] = (df['Date'] - df['Date'].min()).dt.days
df = df.sort_values(by='Date').reset_index(drop=True)
df

Date  Dividends  Stock Splits Ticker  \
0     2000-01-03 05:00:00+00:00        0.0           0.0     TM   
1     2000-01-03 05:00:00+00:00        0.0           0.0     KO   
2     2000-01-03 05:00:00+00:00        0.0           0.0   CSCO   
3     2000-01-03 05:00:00+00:00        0.0           0.0     FL   
4     2000-01-03 05:00:00+00:00        0.0           0.0     HD   
...                         ...        ...           ...    ...   
99784 2023-09-20 04:00:00+00:00        0.0           0.0   TSLA   
99785 2023-09-20 04:00:00+00:00        0.0           0.0  PMMAF   
99786 2023-09-20 04:00:00+00:00        0.0           0.0   NVDA   
99787 2023-09-20 04:00:00+00:00        0.0           0.0   LOGI   
99788 2023-09-20 04:00:00+00:00        0.0           0.0      V   

          Industry_Tag      Country       Close  Year  Month  Day  DayOfWeek  \
0           automotive        japan   72.034958  2000      1    3          0   
1      food & beverage          usa   14.781797  2000      1    3          0   
2           technology          usa   37.496925  2000      1    3          0   
3             footwear          usa    3.957437  2000      1    3          0   
4               retail          usa   40.083256  2000      1    3          0   
...                ...          ...         ...   ...    ...  ...        ...   
99784       automotive          usa  262.589996  2023      9   20          2   
99785          apparel      germany   63.369999  2023      9   20          2   
99786       technology          usa  422.390015  2023      9   20          2   
99787       technology  switzerland   71.800003  2023      9   20          2   
99788          finance          usa  241.860001  2023      9   20          2   

       Day_Number  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
99784        8660  
99785        8660  
99786        8660  
99787        8660  
99788        8660  

[99789 rows x 12 columns]

In [151]:
df['Close_Lag_1'] = df.groupby('Ticker')['Close'].shift(1)
df['Close_Lag_2'] = df.groupby('Ticker')['Close'].shift(2)
df['Close_Lag_3'] = df.groupby('Ticker')['Close'].shift(3)
df['Close_Lag_4'] = df.groupby('Ticker')['Close'].shift(4)
df['Close_Lag_5'] = df.groupby('Ticker')['Close'].shift(5)


# Calculate day differences between current and lagged dates
df['Days_Since_Lag_1'] = df.groupby('Ticker')['Date'].diff(1).dt.days
df['Days_Since_Lag_2'] = df.groupby('Ticker')['Date'].diff(2).dt.days
df['Days_Since_Lag_3'] = df.groupby('Ticker')['Date'].diff(3).dt.days
df['Days_Since_Lag_4'] = df.groupby('Ticker')['Date'].diff(4).dt.days
df['Days_Since_Lag_5'] = df.groupby('Ticker')['Date'].diff(5).dt.days

df[df['Ticker'] == 'FL'][:30]

Date  Dividends  Stock Splits Ticker Industry_Tag  \
3   2000-01-03 05:00:00+00:00        0.0           0.0     FL     footwear   
14  2000-01-04 05:00:00+00:00        0.0           0.0     FL     footwear   
31  2000-01-06 05:00:00+00:00        0.0           0.0     FL     footwear   
63  2000-01-10 05:00:00+00:00        0.0           0.0     FL     footwear   
72  2000-01-11 05:00:00+00:00        0.0           0.0     FL     footwear   
81  2000-01-12 05:00:00+00:00        0.0           0.0     FL     footwear   
106 2000-01-14 05:00:00+00:00        0.0           0.0     FL     footwear   
182 2000-01-25 05:00:00+00:00        0.0           0.0     FL     footwear   
190 2000-01-26 05:00:00+00:00        0.0           0.0     FL     footwear   
208 2000-01-27 05:00:00+00:00        0.0           0.0     FL     footwear   
214 2000-01-28 05:00:00+00:00        0.0           0.0     FL     footwear   
256 2000-02-02 05:00:00+00:00        0.0           0.0     FL     footwear   
260 2000-02-03 05:00:00+00:00        0.0           0.0     FL     footwear   
280 2000-02-04 05:00:00+00:00        0.0           0.0     FL     footwear   
300 2000-02-08 05:00:00+00:00        0.0           0.0     FL     footwear   
334 2000-02-11 05:00:00+00:00        0.0           0.0     FL     footwear   
404 2000-02-18 05:00:00+00:00        0.0           0.0     FL     footwear   
453 2000-02-25 05:00:00+00:00        0.0           0.0     FL     footwear   
504 2000-03-03 05:00:00+00:00        0.0           0.0     FL     footwear   
521 2000-03-07 05:00:00+00:00        0.0           0.0     FL     footwear   
532 2000-03-08 05:00:00+00:00        0.0           0.0     FL     footwear   
544 2000-03-09 05:00:00+00:00        0.0           0.0     FL     footwear   
574 2000-03-14 05:00:00+00:00        0.0           0.0     FL     footwear   
624 2000-03-20 05:00:00+00:00        0.0           0.0     FL     footwear   
648 2000-03-22 05:00:00+00:00        0.0           0.0     FL     footwear   
653 2000-03-23 05:00:00+00:00        0.0           0.0     FL     footwear   
727 2000-03-31 05:00:00+00:00        0.0           0.0     FL     footwear   
731 2000-04-03 04:00:00+00:00        0.0           0.0     FL     footwear   
769 2000-04-06 04:00:00+00:00        0.0           0.0     FL     footwear   
842 2000-04-14 04:00:00+00:00        0.0           0.0     FL     footwear   

    Country     Close  Year  Month  Day  ...  Close_Lag_1  Close_Lag_2  \
3       usa  3.957437  2000      1    3  ...          NaN          NaN   
14      usa  3.886768  2000      1    4  ...     3.957437          NaN   
31      usa  3.922102  2000      1    6  ...     3.886768     3.957437   
63      usa  3.992771  2000      1   10  ...     3.922102     3.886768   
72      usa  3.886768  2000      1   11  ...     3.992771     3.922102   
81      usa  3.851434  2000      1   12  ...     3.886768     3.992771   
106     usa  3.674762  2000      1   14  ...     3.851434     3.886768   
182     usa  3.674762  2000      1   25  ...     3.674762     3.851434   
190     usa  3.639429  2000      1   26  ...     3.674762     3.674762   
208     usa  3.639429  2000      1   27  ...     3.639429     3.674762   
214     usa  3.356755  2000      1   28  ...     3.639429     3.639429   
256     usa  3.639429  2000      2    2  ...     3.356755     3.639429   
260     usa  3.427423  2000      2    3  ...     3.639429     3.356755   
280     usa  3.498093  2000      2    4  ...     3.427423     3.639429   
300     usa  3.427423  2000      2    8  ...     3.498093     3.427423   
334     usa  3.427423  2000      2   11  ...     3.427423     3.498093   
404     usa  3.250752  2000      2   18  ...     3.427423     3.427423   
453     usa  3.109413  2000      2   25  ...     3.250752     3.427423   
504     usa  3.922102  2000      3    3  ...     3.109413     3.250752   
521     usa  3.816098  2000      3    7  ...     3.922102     3.109413   
532     usa  3.957437  2000      3    8  ...     3.8

In [152]:
df['Weighted_Mean_2'] = (
    df['Days_Since_Lag_1'] * df['Close_Lag_1'] + 
    df['Days_Since_Lag_2'] * df['Close_Lag_2']
) / (df['Days_Since_Lag_1'] + df['Days_Since_Lag_2'])

df['Weighted_Mean_5'] = (
    df['Days_Since_Lag_1'] * df['Close_Lag_1'] + 
    df['Days_Since_Lag_2'] * df['Close_Lag_2'] + 
    df['Days_Since_Lag_3'] * df['Close_Lag_3'] + 
    df['Days_Since_Lag_4'] * df['Close_Lag_4'] + 
    df['Days_Since_Lag_5'] * df['Close_Lag_5']
) / (
    df['Days_Since_Lag_1'] + 
    df['Days_Since_Lag_2'] + 
    df['Days_Since_Lag_3'] + 
    df['Days_Since_Lag_4'] + 
    df['Days_Since_Lag_5']
)

df['Weighted_Std_2'] = np.sqrt(
    (df['Days_Since_Lag_1'] * (df['Close_Lag_1'] - df['Weighted_Mean_2'])**2 + 
     df['Days_Since_Lag_2'] * (df['Close_Lag_2'] - df['Weighted_Mean_2'])**2) / 
    (df['Days_Since_Lag_1'] + df['Days_Since_Lag_2'])
)

df['Weighted_Std_5'] = np.sqrt(
    (df['Days_Since_Lag_1'] * (df['Close_Lag_1'] - df['Weighted_Mean_5'])**2 + 
     df['Days_Since_Lag_2'] * (df['Close_Lag_2'] - df['Weighted_Mean_5'])**2 + 
     df['Days_Since_Lag_3'] * (df['Close_Lag_3'] - df['Weighted_Mean_5'])**2 + 
     df['Days_Since_Lag_4'] * (df['Close_Lag_4'] - df['Weighted_Mean_5'])**2 + 
     df['Days_Since_Lag_5'] * (df['Close_Lag_5'] - df['Weighted_Mean_5'])**2) / 
    (df['Days_Since_Lag_1'] + 
     df['Days_Since_Lag_2'] + 
     df['Days_Since_Lag_3'] + 
     df['Days_Since_Lag_4'] + 
     df['Days_Since_Lag_5'])
)

In [ ]:
df[df['Ticker'] == 'FL'][:20]
# df['Dividend_Indicator'] = df['Dividends'].apply(lambda x: 1 if x > 0 else 0)  # 1 if dividend was paid, else 0
# df['Stock_Split_Indicator'] = df['Stock Splits'].apply(lambda x: 1 if x > 0 else 0)  # 1 if a split happened, else 0

 

Add average industry return and avg industry growth yoy
same for country

In [156]:
print(df.dtypes)

Date                     datetime64[ns, UTC]
Dividends                            float64
Stock Splits                         float64
Ticker                                object
Industry_Tag                          object
Country                               object
Close                                float64
Year                                   int32
Month                                  int32
Day                                    int32
DayOfWeek                              int32
Day_Number                             int64
Close_Lag_1                          float64
Close_Lag_2                          float64
Close_Lag_3                          float64
Close_Lag_4                          float64
Close_Lag_5                          float64
Days_Since_Lag_1                     float64
Days_Since_Lag_2                     float64
Days_Since_Lag_3                     float64
Days_Since_Lag_4                     float64
Days_Since_Lag_5                     float64
Weighted_M

In [157]:
numeric_columns = [
    "Dividends",
    "Stock Splits",
    "Close",
    "Year",
    "Month",
    "Day",
    "DayOfWeek",
    "Day_Number",
    "Close_Lag_1",
    "Close_Lag_2",
    "Close_Lag_3",
    "Close_Lag_4",
    "Close_Lag_5",
    "Days_Since_Lag_1",
    "Days_Since_Lag_2",
    "Days_Since_Lag_3",
    "Days_Since_Lag_4",
    "Days_Since_Lag_5",
    "Weighted_Mean_2",
    "Weighted_Mean_5",
    "Weighted_Std_2",
    "Weighted_Std_5"
]

# Categorical columns (nominal / binary / indicators)
categorical_columns = [
    "Date",
    "Ticker",
    "Industry_Tag",
    "Country",
    "Dividend_Indicator",
    "Stock_Split_Indicator"
]

In [ ]:
numeric_preprocessor = Pipeline([
    ("scaler", StandardScaler())
])



split time per ticker

lin reg ridge lasso

random forest, catboost, xgboost

1 classification price up or down

KNN regession

NN